### ISSUE TITLE
Would you mind clarifying how to add additional Roles/Users to the AWS AUTH during EKS provisioning

### ISSUE THREAD

**Issue by vovkanaz** [terraform-aws-modules/terraform-aws-eks#1901]
## Is your request related to a problem? Please describe.
Could not add Roles arn:: to the aws_auth configmap during installation process hence there are no any ability to manage EKS cluster after provisioning
###Unauthorized

## Describe the solution you'd like.
Add ability to manage aws_auth configmap and roles, as it was on a previous versions or maybe provide some information how to figure out this issue related to the current module version 18.7...

## Describe alternatives you've considered.
We've successfully figure this with a help of the local_exec and kubectl_patch, but this approach working well only from local machine not from for example gitlab-runner container. (because there are no any kubectl etc... inside it)

## Additional context
Our team walked through change-log carefully, but still doesn't understood how to manage additional iam users/roles for aws_auth. 

Hopefully, you will help us to find out what is going on and resolve this issue, thanks in advance.

**Comment by stv-io** [comment]
If it's any help, one approach which we're using, using 18.7.2 of the module, and gitlab runners:

 - provision the cluster, generate the kubeconfig using local and `yamlencode`, exposing it as a sensitive output
 - after the apply, capture the kubeconfig - `terraform output -raw kubeconfig` - and export it as a job [artefact](https://docs.gitlab.com/ee/ci/pipelines/job_artifacts.html)
 - in a subsequent job, (same pipeline) manage and overwrite the aws-auth configmap using a `kubectl` container, with the kubeconfig exported from terraform earlier

**Comment by leftyb** [comment]
Hi,
I am having same issue, finally I am going to use similar approach as @stv-io .
I can see that in the past there was a variable that could be set : `map_roles` which somehow it seems that it has been removed.

For now I am creating a template to update the configmap using the outputs of the EKS module.

**Comment by dejwsz** [comment]
Look at my comment there: https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1744#issuecomment-1046810281, maybe it will help a little ;)

**Comment by bryantbiggs** [comment]
I will kindly point you to this issue which would provide better native Terraform support for your request https://github.com/hashicorp/terraform-provider-kubernetes/issues/723

or you can look at a ("hacky") alternative that has been provided in the examples directory https://github.com/terraform-aws-modules/terraform-aws-eks/blob/9a99689cc13147f4afc426b34ba009875a28614e/examples/complete/main.tf#L301-L336

**Comment by aidanmelen** [comment]
I wrote a small module called [eks-auth](https://registry.terraform.io/modules/aidanmelen/eks-auth/aws/latest) to bridge the gap. Here is a code snippet from the complete example:

```hcl
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = ">= 18.0.0"

  cluster_name = var.name
  vpc_id       = module.vpc.vpc_id
  subnet_ids   = module.vpc.private_subnets

  eks_managed_node_groups = {
    foo = {}
  }

  fargate_profiles = {
    bar = {}
  }
}

module "aws_auth" {
  source = "../../"

  eks_aws_auth_configmap_yaml = module.eks.aws_auth_configmap_yaml

  map_roles = [
    {
      rolearn  = "arn:aws:iam::66666666666:role/role1"
      username = "role1"
      groups   = ["system:masters"]
    },
  ]

  map_users = [
    {
      userarn  = "arn:aws:iam::66666666666:user/user1"
      username = "user1"
      groups   = ["system:masters"]
    },
    {
      userarn  = "arn:aws:iam::66666666666:user/user2"
      username = "user2"
      groups   = ["system:masters"]
    },
  ]

  map_accounts = [
    "777777777777",
    "888888888888",
  ]
}
```

The `aws-auth` roles created by `eks_managed_node_groups` and `fargate_profiles` will be used to initialize a new `aws-auth` ConfigMap managed by Terraform.

**Comment by bryantbiggs** [comment]
for anyone following here, check out https://github.com/aws/containers-roadmap/issues/185 - this will solve a number of issues and hopefully get quickly propagated into the Terraform AWS provider (bonus points - someone head over to the provider repo and file a ticket to have day 1 support for it)

**Comment by dmi-clopez** [comment]
What we did to manage `aws-auth` is the following:

We created a local that merges the base `aws-auth` document with our custom changes (credit to @dejwsz for the [snippet](https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1744#issuecomment-1046810281)):


```hcl
  base_auth_configmap = yamldecode(module.eks_cluster.aws_auth_configmap_yaml)

  updated_auth_configmap_data = {
    data = {
      mapRoles = yamlencode(
        distinct(concat(
          yamldecode(local.base_auth_configmap.data.mapRoles), var.map_roles, )
      ))
      mapUsers = yamlencode(var.map_users)
    }
  }

```

Then, we created a Terraform object to manage the `configmap` 

```hcl
resource "kubernetes_config_map" "aws_auth" {

  metadata {
    name      = "aws-auth"
    namespace = "kube-system"
  }

  data = {
    mapAccounts = "[]"
    mapRoles    = local.updated_auth_configmap_data.data.mapRoles
    mapUsers    = local.updated_auth_configmap_data.data.mapUsers
  }
}
```

We tried to use the local-exec solution but it broke our CI/CD workflow (mostly because the `hashicorp/terraform` container doesn't have `bash` installed, and we didn't really want to add a step just to install it), so we came out with this solution.

**Comment by aidanmelen** [comment]
@dmi-clopez are you importing the merged `kubernetes_config_map` resource? Won't terraform fail to apply since the `aws-auth` configMap will already exist in the cluster?

**Comment by dmi-clopez** [comment]
@aidanmelen That's correct for existing clusters. The `aws-auth` configmap does not exist by default for new clusters, so we just create it there.

**Comment by bryantbiggs** [comment]
> @aidanmelen That's correct for existing clusters. The `aws-auth` configmap does not exist by default for new clusters, so we just create it there.

FYI - it will exist on new clusters if using EKS managed node groups or Fargate profiles - those features automatically create the config map and inject their roles into it

**Comment by visla-xugeng** [comment]
I keep receiving this error, when I try to apply the (resource "kubernetes_config_map" "aws_auth") mentioned by @dmi-clopez . Besides this one, when I tried to a new namespace for the eks cluster (resource "kubernetes_namespace"), I got similar timeout error. 

Post "https://11111111111111111.sk1.us-east-1.eks.amazonaws.com/apis/rbac.authorization.k8s.io/v1/clusterrolebindings": dial tcp 10.100.94.191:443: i/o timeout

I cannot access to the eks cluster through kubeconfig, I cannot troubleshoot further. Any thoughts? Thanks,

**Comment by visla-xugeng** [comment]
Hi All, let me re-phrase my issue:

My current configuration can create eks with managed node group successfully. However, I cannot access to the eks cluster, as a result, I am not able to change anything in the eks cluster. For example, I tried to patch the aws-auth configmap with more users and roles, but failed, since I am not able to access to the cluster at all.

@bryantbiggs any idea would be helpful and appreciated. 

`│ Error: Post "https://79EF4ADA33066A605FA13B66222222222.gr7.us-east-1.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps": dial tcp 10.100.69.1:443: i/o timeout
│
│   with module.eks.kubernetes_config_map.aws_auth,
│   on ../../modules/eks_v18/auth-config.tf line 120, in resource "kubernetes_config_map" "aws_auth":
│  120: resource "kubernetes_config_map" "aws_auth" {
│`

`│ Error: Post "https://79EF4ADA33066A605FA13B662F2222222.gr7.us-east-1.eks.amazonaws.com/apis/rbac.authorization.k8s.io/v1/clusterroles": dial tcp 10.100.69.1:443: i/o timeout
│
│   with module.eks.kubernetes_cluster_role.developers,
│   on ../../modules/eks_v18/eks-role-mapping.tf line 3, in resource "kubernetes_cluster_role" "developers":
│    3: resource "kubernetes_cluster_role" "developers" {
│
╵`

**Comment by bryantbiggs** [comment]
the IAM role/user who provisioned the cluster has access to the cluster

**Comment by visla-xugeng** [comment]
> the IAM role/user who provisioned the cluster has access to the cluster

Hi @bryantbiggs The cluster was created by myself, but I cannot connect to it. One thing I want to mention here is that I disabled the "create_iam_role" in the eks module. (I am not sure if the issue is related to the cluster iam role. )

`create_iam_role          = false` 
since I faced the error: Invalid for_each argument when I enable create_iam_role. I have to create the iam role for the eks cluster separately and attached 3 policies to the role.

`
AmazonEKSClusterPolicy
AmazonEKSVPCResourceController
AmazonEKSServicePolicy
`

**Comment by visla-xugeng** [comment]
> the IAM role/user who provisioned the cluster has access to the cluster

Where should I add my account(iam role/user) information to this eks cluster, in the eks module configuration?

**Comment by visla-xugeng** [comment]
@bryantbiggs @dmi-clopez 
I got a wired issue here: I cannot update the aws-auth configmap on my side through `kubernetes_config_map` 
1: the eks cluster will create aws-auth cm when I create an eks cluster with managed-node-groups. All of node-group roles are injected into my aws-auth map. 
below is the aws-auth configmap created by eks
`kubectl -n kube-system get cm aws-auth -oyaml
apiVersion: v1
data:
  mapRoles: |
    - groups:
      - system:bootstrappers
      - system:nodes
      rolearn: arn:aws:iam::111111111111:role/us-east-1-test-default_mng
      username: system:node:{{EC2PrivateDNSName}}
    - groups:
      - system:bootstrappers
      - system:nodes
      rolearn: arn:aws:iam::111111111111:role/us-east-1-test-system_mng
      username: system:node:{{EC2PrivateDNSName}}
    - groups:
      - system:bootstrappers
      - system:nodes
      rolearn: arn:aws:iam::111111111111:role/us-east-1-test-monitoring_mng
      username: system:node:{{EC2PrivateDNSName}}
kind: ConfigMap
metadata:
  creationTimestamp: "2022-03-08T18:43:09Z"
  name: aws-auth
  namespace: kube-system
  resourceVersion: "1150"
  uid: dddddddddddddddddd`


2: I tried @dmi-clopez solution to map other roles and users into aws-auth. However, only users are injected into the aws-auth configmap. `kubernetes_config_map` cannot inject any other new roles into aws-auth, at the same time, the original node-group roles get removed completed. 

Below is what I got after applied the `resource "kubernetes_config_map" "aws_auth"`
`kubectl -n kube-system get cm aws-auth -oyaml
apiVersion: v1
data:
  mapAccounts: '[]'
  mapRoles: |
    - "groups":
      - "system:bootstrappers"
      - "system:nodes"
      "rolearn": null
      "username": "system:node:{{EC2PrivateDNSName}}"
  mapUsers: |
    - "groups":
      - "system:masters"
      "userarn": "arn:aws:iam::1111111:user/user01"
      "username": "user01"
kind: ConfigMap
metadata:
  creationTimestamp: "2022-03-08T05:07:04Z"
  labels:
    app.kubernetes.io/managed-by: Terraform
    terraform.io/module: terraform-aws-modules.eks.aws
  name: aws-auth
  namespace: kube-system
  resourceVersion: "7332"
  uid: xxxxxx`

**Comment by sufyanadam** [comment]
> Previous module versions provided support for managing the aws-auth configmap via the Kubernetes Terraform provider using the now deprecated aws-iam-authenticator; these are no longer included in the module.

This is too bad, it was so nice to be able to provision a cluster and all the team members who should be able to access it in one shot with `terraform apply` and just defining an array of `map_users`.

Please consider putting that back as it was significantly better development experience than what we have to do now with v18+

**Comment by bryantbiggs** [comment]
> > Previous module versions provided support for managing the aws-auth configmap via the Kubernetes Terraform provider using the now deprecated aws-iam-authenticator; these are no longer included in the module.
> 
> 
> 
> This is too bad, it was so nice to be able to provision a cluster and all the team members who should be able to access it in one shot with `terraform apply` and just defining an array of `map_users`.
> 
> 
> 
> Please consider putting that back as it was significantly better development experience than what we have to do now with v18+

https://github.com/aws/containers-roadmap/issues/185

**Comment by gfrid** [comment]
this is very simple, you only add new roles/users that you need, then the patch command adds it to existing config map of EC2 Roles.

locals {

  kubeconfig = yamlencode({
    apiVersion      = "v1"
    kind            = "Config"
    current-context = "terraform"
    clusters = [{
      name = module.eks.cluster_id
      cluster = {
        certificate-authority-data = module.eks.cluster_certificate_authority_data
        server                     = module.eks.cluster_endpoint
      }
    }]
    contexts = [{
      name = "terraform"
      context = {
        cluster = module.eks.cluster_id
        user    = "terraform"
      }
    }]
    users = [{
      name = "terraform"
      user = {
        token = data.aws_eks_cluster_auth.this.token
      }
    }]
  })

  # we have to combine the configmap created by the eks module with the externally created node group/profile sub-modules
  aws_auth_configmap_yaml = <<-EOT
  ${chomp(module.eks.aws_auth_configmap_yaml)}
      - rolearn: arn:aws:iam::**${var.account_id}:role/<your-role>**
        username: Admin
        groups:
          - system:masters
      - rolearn: arn:aws:iam::**${var.account_id}:role/Admin**
        username: Admin
        groups:
          - system:masters
  EOT
}

**Comment by pen-pal** [comment]
@junaid-ali @all I keep getting this error while trying to carry out the upgrade process.
the module I am using also is used to create ingress resources at the same time once the eks cluster is up and running along with externalsecrets manager as well in my case

```
│ Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
│ Error: Failed to get RESTMapper client
│ cannot create discovery client: no client config
│ Error: Failed to get RESTMapper client
│ cannot create discovery client: no client config
│ Error: Failed to get RESTMapper client
│ cannot create discovery client: no client config
│ Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
```

**Comment by aidanmelen** [comment]
I am trying to generate some awareness about a [PR for patch support in the terraform-kubernetes-provider](https://github.com/gavinbunney/terraform-provider-kubectl/issues/64). Please give it a 👍 

With this change, we can apply/patch the `aws-auth` configmap using the `kubernetes_manifest` resource, even if the `aws-auth` configmap already exists.

Here is a example of the terraform that I used to test:

```hcl
provider "kubernetes" {
  config_path    = "~/.kube/config"
  config_context = "docker-desktop"
}

resource "kubernetes_manifest" "aws-auth-configmap" {
  manifest = yamldecode(
    <<-EOT
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: aws-auth
      namespace: kube-system
    data:
      mapRoles: |
        - rolearn: arn:aws:iam::111111111111:role/DemoEKS-NodeInstanceRole
          username: system:node:{{EC2PrivateDNSName}}
          groups:
            - system:bootstrappers
            - system:nodes
        - rolearn: arn:aws:iam::111111111111:role/TeamRole
          username: TeamRole
          groups:
          - system:masters
      mapUsers: |
        - userarn: arn:aws:iam::111111111111:user/sukumar-test-test
          username: sukumar
          groups:
            - system:masters
    EOT
  )

  field_manager {
    force_conflicts = true 
  }
}
```

**Comment by antonbabenko** [comment]
This issue has been resolved in version 18.20.0 :tada:

**Comment by rajeshgopinath** [comment]
> What we did to manage `aws-auth` is the following:
> 
> We created a local that merges the base `aws-auth` document with our custom changes (credit to @dejwsz for the [snippet](https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1744#issuecomment-1046810281)):
> 
> ```terraform
>   base_auth_configmap = yamldecode(module.eks_cluster.aws_auth_configmap_yaml)
> 
>   updated_auth_configmap_data = {
>     data = {
>       mapRoles = yamlencode(
>         distinct(concat(
>           yamldecode(local.base_auth_configmap.data.mapRoles), var.map_roles, )
>       ))
>       mapUsers = yamlencode(var.map_users)
>     }
>   }
> ```
> 
> Then, we created a Terraform object to manage the `configmap`
> 
> ```terraform
> resource "kubernetes_config_map" "aws_auth" {
> 
>   metadata {
>     name      = "aws-auth"
>     namespace = "kube-system"
>   }
> 
>   data = {
>     mapAccounts = "[]"
>     mapRoles    = local.updated_auth_configmap_data.data.mapRoles
>     mapUsers    = local.updated_auth_configmap_data.data.mapUsers
>   }
> }
> ```
> 
> We tried to use the local-exec solution but it broke our CI/CD workflow (mostly because the `hashicorp/terraform` container doesn't have `bash` installed, and we didn't really want to add a step just to install it), so we came out with this solution.

Will this update an existing config map or it expects aws-auth configmap doesn't exist?
